### PR TITLE
Allow forms to be autofilled with queries in urls

### LINKED ISSUE #429: Allow forms to be prefilled
Sort of like what Reddit does ([example](https://www.reddit.com/message/compose?to=FlairHQ&subject=Prefilled%20forms%21&message=Whoa%2C%20how%20did%20this%20text%20get%20here?)), where form fields can be prefilled with a url query.

### LINKED ISSUE #433: Flairs don't appear in the Apply For Flair menu
[Pretty much what it says on the tin.](https://i.gyazo.com/00967ccd9c54e890470a21c7a752c8d3.png) (This is a 2.3 issue, it works fine in 2.2.)

### PR DESCRIPTION
**Pull request by not-an-aardvark**
Addresses #429, fixes #433.

Example urls (click while your local server is running):

* [Adding a trade](http://localhost:1337/?refUrl=https://reddit.com/r/pokemontrades/comments/blah/blah/blah&type=event&user2=Satan&gave=My+Soul&got=Cookies&notes=They+were+chocolate+chip+cookies%2E&privatenotes=I+actually+gave+someone+else%27s+soul+instead%2E)
* [Banning someone](http://localhost:1337/mod/banuser?username=actually_an_aardvark&banNote=Pretending+to+be+an+aardvark&banMessage=You+have+been+banned+because+the+mods+are+evil%2E&banlistEntry=Breaking+the+rules%2C+or+something+to+that+effect)

This automatically prefills the forms, but it does not automatically press the submit button, for obvious reasons. (We don't want someone to be able to indirectly ban a user by getting a mod to click a link, for example.)

In theory, we could make it work automatically if we required that a secure token be included in the query. For example, FlairHQ could send a user a PM:

`A user added a reference for their trade with you. Click [here](hq.porygon.co/some_url?token=07a369a29ad9574947c99601bdcf293d17112f85) to add it to your reference as well.`

...and it could be a one-click action.

### CODE REVIEW

**Comment by YaManicKill** [comment]
Suggestion: can you add it server-side as well, so that we don't get further away from everything working without JavaScript

Add it as the value to the input fields on the server.

**Comment by not-an-aardvark** [comment]
> Add it as the value to the input fields on the server.

I'm not sure I understand what this means. Since the form itself and the submit action are handled separately by the server, I don't know which input fields you're referring to.

File: api/policies/sessionAuth.js [hunk 1]
```diff
@@ -21,10 +21,6 @@ module.exports = function(req, res, next) {
   // User is allowed, proceed to the next policy,
   // or if this is the last policy, the controller
   if (req.user || (req.isAuthenticated && req.isAuthenticated())) {
-    //Redirect mods to the modauth page if they only have normal user scope.
-    if (req.user.isMod && req.session.state.substr(-9) === '_modlogin') {
-      return res.redirect('/auth/modauth');
-    }
```
Comment (not-an-aardvark) [inline api/policies/sessionAuth.js]: This isn't needed anymore because we now check whether a user is a mod before we log them in.

**Comment by YaManicKill** [comment]
Well, I mean we have the form in an ejs file on the server. If we pass the values from url into it as a local, we can pre-render the values into the form inputs, so that for people without JavaScript, this feature still works (once everything else works, obviously it doesn't work just now because the actual page won't load fully, but I'm working on that.

So, for example: home/index.ejs:28

<input type="url" class="form-control" ng-model="addInfo.refUrl" />

change it to this

<input type="url" class="form-control" ng-model="addInfo.refUrl" value="<%= formValues.url %>" />

Obviously, doesn't need to be exactly that, but you get my point now?

I'm currently trying to figure out a nice way to render certain things server-side and then still let Angular take them over. But this should work this way.

File: assets/views/home/flairText.ejs [hunk 2]
```diff
@@ -50,7 +50,7 @@
             </div>
 
             <div class="form-group col-sm-12" ng-repeat="game in user.flairGames | orderBy:'createdAt' track by $index">
-              <label class="col-md-3 control-label">IGN:*</label>
+              <label class="col-md-3 control-label" title="In-Game Name (i.e. the name of your character)">IGN:*</label>
```
Comment (not-an-aardvark) [inline assets/views/home/flairText.ejs]: This isn't related, but it didn't fit anywhere else so I dropped it here.

File: assets/views/home/header.ejs [hunk 3]
```diff
@@ -55,7 +55,7 @@
 
       <ul class="nav navbar-nav hidden" ng-hide="user" ng-class="loaded ? '' : 'hidden'">
         <li><a href="/info">Information</a></li>
-        <li><a href="/auth/reddit/">Login with Reddit</a></li>
+        <li><a href="/auth/reddit{{querystring}}">Login with Reddit</a></li>
```
Comment (YaManicKill) [inline assets/views/home/header.ejs]: This won't change while we are on the page, will it? It's based on the query string from the url, so can you render it serverside instead? I've just changed the header to work mostly without javascript (if you ignore the dropdowns).

**Comment by not-an-aardvark** [comment]
Alright, I added the default query parameters serverside as well. Note that they don't actually do anything at the moment, because the `ng-model` attached to them just clears the values anyway, and then sets them later on in the controllers. But they should now have default values if javascript is disabled.

File: assets/userCtrl.js [hunk 4]
```diff
@@ -475,4 +478,18 @@ module.exports = function ($scope, $filter, $location, UserFactory) {
...
+  io.socket.get('/csrfToken', function (token) {
+    $scope._csrf = token._csrf;
+  });
+
+  /* Include the csrf token in all POST requests.
+   * (As far as I can tell, there's no better way to do this aside from manually adding the token parameter to each request.)*/
+  io.socket.post = function (url, data, callback) {
+    data._csrf = $scope._csrf;
+    io.socket.request({method: 'post', url: url, params: data}, callback);
+  };
```
Comment (not-an-aardvark) [inline assets/userCtrl.js]: This is not an ideal solution, but it's better than having tons of duplicate code, and I couldn't find a better one.

Comment (YaManicKill) [inline assets/userCtrl.js]: Maybe add an angular service to override this rather than putting it in this controller?

**Comment by YaManicKill** [comment]
Just the 1 thing, then I'm happy to merge.

File: assets/views/layout.ejs [hunk 5]
```diff
@@ -62,7 +62,7 @@
     </div>
 
     <!--SCRIPTS-->
-    <script src="/js/app.js"></script>
+    <script id="app" src="/js/app.js" _csrf="<%= _csrf %>"></script>
```
Comment (YaManicKill) [inline assets/views/layout.ejs]: Instead of doing it this way, set it on the body tag and add an ng-init

Comment (not-an-aardvark) [inline assets/views/layout.ejs]: Won't ng-init fail since we don't have access to `$scope` in app.js?

File: assets/search/main.ejs [hunk 6]
```diff
@@ -16,16 +16,16 @@
...
             <a ng-repeat="result in search.results track by $index"
                ng-href="{{search.linkAddress(result)}}"
                ng-if="result !== null"
                class="list-group-item">
               <%- partial(searchType + '/result.ejs') %>
             </a>
+            <li class="list-group-item"
+                ng-if="search.inProgress">
+              Searching...
+            </li>
```
Comment (not-an-aardvark) [inline assets/search/main.ejs]: Another unrelated fix: I moved the 'Searching' message to be below the results (in the same place as the 'get more results' button), rather than above. Now the results don't jump down when you try to get more.

File: assets/views/layout.ejs (hunk 5 above)
Comment (not-an-aardvark) [inline assets/views/layout.ejs]: I don't think ng-init will work, because we want the service function to be created before the controllers are created, but ng-init gets called after the controllers are created.
